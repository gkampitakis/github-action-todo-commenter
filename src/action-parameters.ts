import { getInput } from '@actions/core';
import { Context } from '@actions/github/lib/context';

import { GetActionParams, GetInputParams } from './types';

export function getInputs(): GetInputParams {
  const tags = getInput('tags') || 'TODO:,FIXME:,BUG:';
  const reviewMsg = getInput('review-message');
  const ignoreFilesPattern = getInput('ignore-pattern');
  const token = getInput('github-token', { required: true });
  const commentTitle = getInput('comment-title') || 'Todo Commenter';

  return {
    tags: tags.split(','),
    reviewMsg,
    token,
    ignoreFilesPattern,
    commentTitle
  };
}

export function getActionParameters(ctx: Context): GetActionParams {
  const {
    actor,
    repo: { owner, repo },
    eventName,
    payload: { pull_request }
  } = ctx;
  const base = pull_request?.base?.ref;
  const head = pull_request?.head?.ref;

  console.log(eventName);

  if (eventName !== 'pull_request') {
    throw new Error('Action only supports pull requests');
  }

  if (!pull_request?.number) {
    throw new Error('Action cannot identify pull request number');
  }

  return {
    actor,
    owner,
    repo,
    base,
    head,
    prNumber: pull_request?.number
  };
}
