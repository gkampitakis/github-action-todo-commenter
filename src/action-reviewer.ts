import { info } from '@actions/core';
import { ActionReviewerOptions, Octokit } from './types';

export class ActionReviewer {
  private octokit: Octokit;
  private options: ActionReviewerOptions;

  constructor(octokit: Octokit, options: ActionReviewerOptions) {
    this.octokit = octokit;
    this.options = options;
  }

  public async createReview(body: string) {
    const { id, body: oldReviewBody } = await this.reviewExists();

    if (oldReviewBody === body) {
      info('No new changes detected since ');
      return;
    }

    if (id) {
      await this.deleteReview(id);
    }

    await this.octokit.rest.issues.createComment({
      owner: this.options.owner,
      repo: this.options.repo,
      issue_number: this.options.prNumber,
      body
    });

    return;
  }

  public async singleCommentReview(
    body: string,
    { line, path }: { line: number; path: string }
  ) {
    return this.octokit.rest.pulls.createReviewComment({
      owner: this.options.owner,
      repo: this.options.repo,
      pull_number: this.options.prNumber,
      commit_id: this.options.commitId,
      position: line,
      body,
      path,
      line
    });
  }
// test
  public async deleteReview(id: number) {
    return this.octokit.rest.issues.deleteComment({
      owner: this.options.owner,
      repo: this.options.repo,
      comment_id: id
    });
  }

  public async reviewExists(): Promise<{ id?: number; body?: string }> {
    const { data: allReviews } = await this.octokit.rest.issues.listComments({
      owner: this.options.owner,
      repo: this.options.repo,
      issue_number: this.options.prNumber
    });

    if (allReviews.length === 0) {
      return {};
    }

    const review = allReviews.find(item => {
      return (
        item.user?.login === 'github-actions[bot]' &&
        item.body?.startsWith(this.options.commentIdentifier)
      );
    });

    if (!review || !review.body) {
      return {};
    }

    // This removes the first line containing the comment identifier
    let index = review.body.match(/[\r\n]+/)?.index;
    index = index ? index + 1 : 0;

    return { id: review?.id, body: review.body.substr(index) };
  }
}
