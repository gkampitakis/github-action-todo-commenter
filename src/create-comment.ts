import { getOctokit } from '@actions/github';
type CreateCommentParams = {
  token: string;
  owner: string;
  repo: string;
  prNumber: number;
};

export async function createComment(
  params: CreateCommentParams,
  body: string,
  block: boolean
) {
  const { token, owner, repo, prNumber } = params;
  const octokit = getOctokit(token);

  const { data: allComments } = await octokit.rest.issues.listComments({
    owner,
    repo,
    issue_number: prNumber
  });

  const comment = allComments.find(item => {
    return (
      item.user?.login === 'github-actions[bot]' &&
      item.body?.startsWith('## Todo Commenter\n')
    );
  });

  if (comment) {
    return await octokit.rest.pulls.updateReviewComment({
      owner,
      repo,
      body,
      comment_id: comment.id,
      pull_number: prNumber
    });
  }

  return octokit.rest.pulls.createReview({
    owner,
    repo,
    event: block ? 'REQUEST_CHANGES' : 'COMMENT',
    pull_number: prNumber,
    body
  });
}
