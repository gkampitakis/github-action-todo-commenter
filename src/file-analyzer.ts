import { createReadStream } from 'fs';
import { createInterface } from 'readline';
import path from 'path';

type FileAnalyzerResult = {
  file: string;
  comments: Comments;
}
type Comments = Record<string, string[]>;
type EnhancedTag = { tag: string, regex: RegExp };

export async function fileAnalyzer(filePaths: string[], tags: string[]): Promise<FileAnalyzerResult[]> {
  const enhancedTags = tags.map(tag => ({ tag, regex: new RegExp(tag + '(.*)') }));
  const ignoreMinSize = Math.min(...tags.map(t => t.length));
  const result: FileAnalyzerResult[] = [];

  for (const filePath of filePaths) {
    const absolutePath = path.join(process.cwd(), filePath);
    const comments = await readFile(absolutePath, enhancedTags, ignoreMinSize);

    result.push({
      file: filePath,
      comments
    });
  }

  return result;
}

async function readFile(filepath: string, tags: EnhancedTag[], ignoreMinSize: number): Promise<Comments> {
  const comments: Comments = {};
  const stream = createReadStream(filepath, 'utf-8');
  const rl = createInterface({
    input: stream,
    crlfDelay: Infinity
  });

  for await (const line of rl) {
    if (ignoreMinSize > line.length) continue;

    for (let { tag, regex } of tags) {
      const matched = line.match(regex);

      if (!matched) continue;

      const comment = matched[1].trim();

      if (comments[tag]) {
        comments[tag].push(comment);
        continue;
      }

      comments[tag] = [comment];
    }
  }

  return comments;
}
