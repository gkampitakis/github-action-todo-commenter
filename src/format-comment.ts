import { FileAnalyzerResults } from './types';

export function formatComment(
  analyzedCommentsPerFile: FileAnalyzerResults,
  { actor, reviewMsg }: { actor?: string; reviewMsg?: string }
) {
  let message = '## Todo Commenter\n';

  analyzedCommentsPerFile.forEach(({ comments, file }) => {
    message += `<details>\n`;
    message += `<summary><strong>${file}</strong></summary>\n`;

    Object.entries(comments).forEach(([key, items]) => {
      message += `\n##### ${key}\n\n`;

      items.forEach(item => {
        if (item === '') {
          message += `- [ ] <space>\n`;
          return;
        }
        message += `- [ ] ${item}\n`;
      });
    });

    message += `</details>\n\n`;
  });

  if (reviewMsg) {
    message += '---\n';
    message += `@${actor} ${reviewMsg}`;
  }

  return message;
}
