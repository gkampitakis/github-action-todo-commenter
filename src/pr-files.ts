import { GetFilesParams } from './types';
import { exec as cbExec } from 'child_process';
import { promisify } from 'util';
import { warning } from '@actions/core';

const exec = promisify(cbExec);

export async function getFiles({
  octokit,
  owner,
  repo,
  prNumber,
  base,
  head,
  ignoreFilesPattern
}: GetFilesParams) {
  const { data: prFiles } = await octokit.rest.pulls.listFiles({
    owner,
    repo,
    pull_number: prNumber,
    mediaType: {
      format: 'diff'
    }
  });

  // Filter out submodules
  const untracked = ['removed', 'unchanged'];
  let matcher = (file: typeof prFiles[number]) =>
    !untracked.includes(file.status) &&
    Boolean(file.additions) &&
    Boolean(file.blob_url);

  if (ignoreFilesPattern) {
    const regex = new RegExp(ignoreFilesPattern);
    matcher = (file: typeof prFiles[number]) =>
      !untracked.includes(file.status) &&
      Boolean(file.blob_url) &&
      Boolean(file.additions) &&
      !file.filename.match(regex);
  }

  return Promise.all(
    prFiles.filter(matcher).map(async ({ filename, patch }) => ({
      filename,
      // Noticed on big diffs the api doesn't return a patch string
      // Could not find documentation around size limit on `patch` field
      patch: patch ? patch : await diff(base, head, filename)
    }))
  );
}

async function diff(
  base: string,
  head: string,
  filename: string
): Promise<string> {
  try {
    const { stdout, stderr } = await exec(
      `git diff origin/${base}...origin/${head} -- ${filename}`
    );
    if (stderr) {
      warning(stderr);
      return '';
    }

    return stdout;
  } catch (err: any) {
    warning(err);
    return '';
  }
}
