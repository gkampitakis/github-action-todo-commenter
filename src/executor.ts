import { context, getOctokit } from '@actions/github';
import { getInput, setFailed, info } from '@actions/core';
import { fileAnalyzer } from './file-analyzer';
import { formatComment } from './format-comment';
import { ActionReviewer } from './action-reviewer';
import { Context } from '@actions/github/lib/context';
import { Octokit } from './types';

export async function run() {
  try {
    const { tags, reviewMsg, token, ignoreFilesPattern } = getInputs();
    const { actor, owner, prNumber, repo } = getActionParameters(context);
    const octokit = getOctokit(token);

    const files = await getFiles({
      octokit,
      repo,
      owner,
      prNumber,
      ignoreFilesPattern
    });
    const analyzedComments = await fileAnalyzer(files, tags);
    const actionReviewer = new ActionReviewer({
      owner,
      repo,
      octokit,
      prNumber
    });

    if (analyzedComments.length === 0) {
      const { id } = await actionReviewer.reviewExists();
      if (id) {
        await actionReviewer.deleteReview(id);
      }

      info('No tags found in pr ðŸ‘€');
      return;
    }

    const comment = formatComment(analyzedComments, { actor, reviewMsg });

    await actionReviewer.createReview(comment);
  } catch (error: any) {
    setFailed(error.message);
  }
}

function getInputs(): {
  token: string;
  reviewMsg: string;
  ignoreFilesPattern: string;
  tags: string[];
} {
  const tags = getInput('tags') || 'TODO:,FIXME:,BUG:';
  const reviewMsg = getInput('review-msg');
  const ignoreFilesPattern = getInput('ignore-pattern');
  const token = getInput('github-token') || process.env.GITHUB_TOKEN || '';

  if (token === '') {
    throw new Error(`Action needs 'GITHUB_TOKEN' in order to work correctly`);
  }

  return {
    tags: tags.split(','),
    reviewMsg,
    token,
    ignoreFilesPattern
  };
}

function getActionParameters(ctx: Context): {
  actor: string;
  owner: string;
  repo: string;
  prNumber: number;
} {
  const {
    actor,
    repo: { owner, repo },
    eventName,
    payload: { pull_request }
  } = ctx;

  if (eventName !== 'pull_request') {
    throw new Error('Action only supports pull requests');
  }

  if (!pull_request?.number) {
    throw new Error('Action cannot identify pull request number');
  }

  return {
    actor,
    owner,
    repo,
    prNumber: pull_request?.number
  };
}

async function getFiles({
  octokit,
  owner,
  repo,
  prNumber,
  ignoreFilesPattern
}: {
  octokit: Octokit;
  owner: string;
  repo: string;
  prNumber: number;
  ignoreFilesPattern: string;
}): Promise<string[]> {
  const { data: prFiles } = await octokit.rest.pulls.listFiles({
    owner,
    repo,
    pull_number: prNumber
  });
  const untracked = ['removed', 'unchanged'];
  let matcher = (file: typeof prFiles[number]) =>
    !untracked.includes(file.status);

  if (ignoreFilesPattern) {
    const regex = new RegExp(ignoreFilesPattern);
    matcher = (file: typeof prFiles[number]) =>
      !untracked.includes(file.status) && !file.filename.match(regex);
  }

  return prFiles.filter(matcher).map(f => f.filename);
}
