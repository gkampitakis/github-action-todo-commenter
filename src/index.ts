import { getInput } from '@actions/core';
import { executor } from './executor';

const tags = ['TODO', 'FIXME'];
const customTags = getInput('tags').split(',');

console.log(tags, customTags);


executor(customTags || tags);
