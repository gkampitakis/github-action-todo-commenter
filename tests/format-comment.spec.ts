import { formatComment } from '../src/format-comment';

const analyzedComments = [
  {
    comments: {
      'FIXME:': [':add implementation'],
      NOTE: ["Please don't forget review", 'another note'],
      'tODo:': ['this should present']
    },
    file: './tests/mockFiles/mockFile0.js'
  },
  {
    comments: {
      'FIXME:': [':add implementation'],
      NOTE: ["Please don't forget review", 'another note'],
      'tODo:': ['this should present']
    },
    file: './tests/mockFiles/mockFile2.js'
  }
];

describe('FormatComment', () => {
  it('should format comment consistently', () => {
    const comment = formatComment(analyzedComments, {});

    expect(comment).toMatchSnapshot();
  });

  it('should tag actor with reviewMsg', () => {
    const comment = formatComment(analyzedComments, {
      actor: 'mock-actor',
      reviewMsg: 'please review'
    });

    expect(comment).toMatchSnapshot();
  });
});
