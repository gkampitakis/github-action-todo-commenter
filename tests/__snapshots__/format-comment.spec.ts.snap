// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormatComment should format comment consistently 1`] = `
"## Todo Commenter
<details>
<summary><strong>./tests/mockFiles/mockFile0.js</strong></summary>

##### FIXME:

- [ ] :add implementation

##### NOTE

- [ ] Please don't forget review
- [ ] another note

##### tODo:

- [ ] this should present
</details>

<details>
<summary><strong>./tests/mockFiles/mockFile2.js</strong></summary>

##### FIXME:

- [ ] :add implementation

##### NOTE

- [ ] Please don't forget review
- [ ] another note

##### tODo:

- [ ] this should present
</details>

"
`;

exports[`FormatComment should tag actor with reviewMsg 1`] = `
"## Todo Commenter
<details>
<summary><strong>./tests/mockFiles/mockFile0.js</strong></summary>

##### FIXME:

- [ ] :add implementation

##### NOTE

- [ ] Please don't forget review
- [ ] another note

##### tODo:

- [ ] this should present
</details>

<details>
<summary><strong>./tests/mockFiles/mockFile2.js</strong></summary>

##### FIXME:

- [ ] :add implementation

##### NOTE

- [ ] Please don't forget review
- [ ] another note

##### tODo:

- [ ] this should present
</details>

---
@mock-actor please review"
`;
